src/Launcher.vala

```diff
@@ -33,9 +33,36 @@ public class Dock.Launcher : Gtk.Button {
         };
         image.get_style_context ().add_provider (css_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
 
-        child = image;
+        // Needed to work around DnD bug where it
+        // would stop working once the button got clicked
+        var box = new Gtk.Box (VERTICAL, 0);
+        box.append (image);
+
+        child = box;
         tooltip_text = app_info.get_display_name ();
 
+        var drag_source = new Gtk.DragSource () {
+            actions = MOVE
+        };
+        box.add_controller (drag_source);
+
+        drag_source.prepare.connect (() => {
+            var val = Value (typeof (Launcher));
+            val.set_object (this);
+            var content_provider = new Gdk.ContentProvider.for_value (val);
+            return content_provider;
+        });
```

Comment:
```suggestion
         var drag_value = Value (typeof (Launcher));
        drag_value.set_object (this);

        var drag_source = new Gtk.DragSource () {
            actions = MOVE,
            content = new Gdk.ContentProvider.for_value (drag_value)
        };
        box.add_controller (drag_source);
```

Since the drag content is always the launcher itself we can do this, instead of connecting to the `prepare` method.